// angularfire2/__ivy_ngcc__/database-deprecated/es2015/database.js
import 'firebase/database';
import { Injectable } from '@angular/core';
import { FirebaseApp } from 'angularfire2';
import { FirebaseListFactory } from './firebase_list_factory';
import { FirebaseObjectFactory } from './firebase_object_factory';
import * as utils from './utils';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'angularfire2';
export class AngularFireDatabase {
    constructor(app) {
        this.app = app;
        this.database = app.database();
    }
    list(pathOrRef, opts) {
        const ref = utils.getRef(this.app, pathOrRef);
        return FirebaseListFactory(ref, opts);
    }
    object(pathOrRef, opts) {
        const ref = utils.getRef(this.app, pathOrRef);
        return FirebaseObjectFactory(ref, opts);
    }
}
AngularFireDatabase.ɵfac = function AngularFireDatabase_Factory(t) { return new (t || AngularFireDatabase)(ɵngcc0.ɵɵinject(ɵngcc1.FirebaseApp)); };
AngularFireDatabase.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: AngularFireDatabase, factory: AngularFireDatabase.ɵfac });
AngularFireDatabase.ctorParameters = () => [
    { type: FirebaseApp, },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireDatabase, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.FirebaseApp }]; }, null); })();
