// angularfire2/__ivy_ngcc__/database/es2015/database.module.js
import { NgModule } from '@angular/core';
import 'firebase/database';
import { AngularFireModule, FirebaseApp } from 'angularfire2';
import { AngularFireDatabase } from './database';
import * as ɵngcc0 from '@angular/core';
export function _getAngularFireDatabase(app) {
    return new AngularFireDatabase(app);
}
export const AngularFireDatabaseProvider = {
    provide: AngularFireDatabase,
    useFactory: _getAngularFireDatabase,
    deps: [FirebaseApp]
};
export const DATABASE_PROVIDERS = [
    AngularFireDatabaseProvider,
];
export class AngularFireDatabaseModule {
}
AngularFireDatabaseModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularFireDatabaseModule });
AngularFireDatabaseModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularFireDatabaseModule_Factory(t) { return new (t || AngularFireDatabaseModule)(); }, providers: [DATABASE_PROVIDERS], imports: [[AngularFireModule]] });
AngularFireDatabaseModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularFireDatabaseModule, { imports: function () { return [AngularFireModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireDatabaseModule, [{
        type: NgModule,
        args: [{
                imports: [AngularFireModule],
                providers: [DATABASE_PROVIDERS]
            }]
    }], null, null); })();
