// angularfire2/__ivy_ngcc__/storage/es2015/storage.module.js
import { NgModule } from '@angular/core';
import 'firebase/storage';
import { AngularFireModule, FirebaseApp } from 'angularfire2';
import { AngularFireStorage } from './storage';
import * as ɵngcc0 from '@angular/core';
export function _getAngularFireStorage(app) {
    return new AngularFireStorage(app);
}
export const AngularFireStorageProvider = {
    provide: AngularFireStorage,
    useFactory: _getAngularFireStorage,
    deps: [FirebaseApp]
};
export const STORAGE_PROVIDERS = [
    AngularFireStorageProvider,
];
export class AngularFireStorageModule {
}
AngularFireStorageModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularFireStorageModule });
AngularFireStorageModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularFireStorageModule_Factory(t) { return new (t || AngularFireStorageModule)(); }, providers: [STORAGE_PROVIDERS], imports: [[AngularFireModule]] });
AngularFireStorageModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularFireStorageModule, { imports: function () { return [AngularFireModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireStorageModule, [{
        type: NgModule,
        args: [{
                imports: [AngularFireModule],
                providers: [STORAGE_PROVIDERS]
            }]
    }], null, null); })();
