// angularfire2/__ivy_ngcc__/storage/es2015/storage.js
import { Injectable } from '@angular/core';
import { FirebaseApp } from 'angularfire2';
import { createStorageRef } from './ref';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'angularfire2';
export class AngularFireStorage {
    constructor(app) {
        this.app = app;
        this.storage = app.storage();
    }
    ref(path) {
        return createStorageRef(this.storage.ref(path));
    }
    upload(path, data, metadata) {
        const storageRef = this.storage.ref(path);
        const ref = createStorageRef(storageRef);
        return ref.put(data, metadata);
    }
}
AngularFireStorage.ɵfac = function AngularFireStorage_Factory(t) { return new (t || AngularFireStorage)(ɵngcc0.ɵɵinject(ɵngcc1.FirebaseApp)); };
AngularFireStorage.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: AngularFireStorage, factory: AngularFireStorage.ɵfac });
AngularFireStorage.ctorParameters = () => [
    { type: FirebaseApp, },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireStorage, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.FirebaseApp }]; }, null); })();
