// angularfire2/__ivy_ngcc__/auth/es2015/auth.module.js
import { NgModule } from '@angular/core';
import 'firebase/auth';
import { FirebaseApp, AngularFireModule } from 'angularfire2';
import { AngularFireAuth } from './auth';
import * as ɵngcc0 from '@angular/core';
export function _getAngularFireAuth(app) {
    return new AngularFireAuth(app);
}
export const AngularFireAuthProvider = {
    provide: AngularFireAuth,
    useFactory: _getAngularFireAuth,
    deps: [FirebaseApp]
};
export const AUTH_PROVIDERS = [
    AngularFireAuthProvider,
];
export class AngularFireAuthModule {
}
AngularFireAuthModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: AngularFireAuthModule });
AngularFireAuthModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function AngularFireAuthModule_Factory(t) { return new (t || AngularFireAuthModule)(); }, providers: [AUTH_PROVIDERS], imports: [[AngularFireModule]] });
AngularFireAuthModule.ctorParameters = () => [];
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(AngularFireAuthModule, { imports: function () { return [AngularFireModule]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireAuthModule, [{
        type: NgModule,
        args: [{
                imports: [AngularFireModule],
                providers: [AUTH_PROVIDERS]
            }]
    }], null, null); })();
