// angularfire2/__ivy_ngcc__/auth/es2015/auth.js
import 'firebase/auth';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { observeOn } from 'rxjs/operator/observeOn';
import { FirebaseApp, ZoneScheduler } from 'angularfire2';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/fromPromise';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from 'angularfire2';
export class AngularFireAuth {
    constructor(app) {
        this.app = app;
        this.auth = app.auth();
        const authState$ = new Observable(subscriber => {
            const unsubscribe = this.auth.onAuthStateChanged(subscriber);
            return { unsubscribe };
        });
        this.authState = observeOn.call(authState$, new ZoneScheduler(Zone.current));
        const idToken$ = new Observable(subscriber => {
            const unsubscribe = this.auth.onIdTokenChanged(subscriber);
            return { unsubscribe };
        }).switchMap(user => {
            return user ? Observable.fromPromise(user.getIdToken()) : Observable.of(null);
        });
        this.idToken = observeOn.call(idToken$, new ZoneScheduler(Zone.current));
    }
}
AngularFireAuth.ɵfac = function AngularFireAuth_Factory(t) { return new (t || AngularFireAuth)(ɵngcc0.ɵɵinject(ɵngcc1.FirebaseApp)); };
AngularFireAuth.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: AngularFireAuth, factory: AngularFireAuth.ɵfac });
AngularFireAuth.ctorParameters = () => [
    { type: FirebaseApp, },
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AngularFireAuth, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.FirebaseApp }]; }, null); })();
